### PR TITLE
Add the ability to compute dependency versions (in a monorepo situation)

### PR DESCRIPTION
**Pull request by chrisbetz**
In a monorepository, I'd like to share the same version (derived from git tags) in all projects, i.e. apps and libs respectively. Therefore, I not only need to compute the project version, but also compute the version of the dependencies in a given project.

Using the new middleware `leiningen.v/dependency-version-from-scm`, you can achive this by a nil-dependency version, just like managed-dependencies are handling this.

I wasn't able to get the tests up and running (before my change), but I got an example repo here, if you'd like to check: [example of my lein-monolith fork](https://github.com/chrisbetz/lein-monolith/tree/master/example). Just run `lein monolith each do clean, install` in the example-folder to check.

I'd really much appreciate any comments or a merge :)

Cheers,

Chris

### CODE REVIEW

**Comment by chrisbetz** [comment]
Ah, and I included the change from @matthiasn , because fixing the README was on my list also ;)

File: project.clj [hunk 1]
```diff
@@ -1,4 +1,4 @@
-(defproject com.roomkey/lein-v :lein-v
+(defproject com.roomkey/lein-v "0.0.0"
```
Comment (cch1) [inline project.clj]: Why is this required?  Last I checked, `lein-v` was self-aware...  but I am aware of issues when backwards compatibility is not provided.  Can you fill me in on your specific issue?

File: src/leiningen/v.clj [hunk 2]
```diff
@@ -103,15 +103,34 @@
...
-       (assoc-in [:manifest vk] v))))
+        (assoc-in [:version] v)
+        (assoc-in [:manifest vk] v))))
+
+(defn- update-dependency [v d]
+  (if-not (second d)
+    (assoc d 1 v)
+    d))
+
+; => (update-dependency "42" ["abc" "1.0"])
```
Comment (cch1) [inline src/leiningen/v.clj]: For consistency sake, please keep examples in README.  I'm looking forward to clojure.spec for some documentation capabilities, but until then I don't want to have to worry about keeping comments updated as code evolves.

File: src/leiningen/v.clj [hunk 3]
```diff
@@ -103,15 +103,34 @@
...
     (-> project
-        (assoc-in ,, [:workspace] wss)
-        (assoc-in ,, [:manifest "Workspace-Description"] (:describe wss))
-        (assoc-in ,, [:manifest "Workspace-Tracking-Status"] (string/join " || " (get-in wss [:status :tracking])))
-        (assoc-in ,, [:manifest "Workspace-File-Status"] (string/join " || " (get-in wss [:status :files]))))
+        (assoc-in,, [:workspace] wss)
+        (assoc-in,, [:manifest "Workspace-Description"] (:describe wss))
+        (assoc-in,, [:manifest "Workspace-Tracking-Status"] (string/join " || " (get-in wss [:status :tracking])))
+        (assoc-in,, [:manifest "Workspace-File-Status"] (string/join " || " (get-in wss [:status :files]))))
     project))
```
Comment (cch1) [inline src/leiningen/v.clj]: I really like this PR -it turns out that I have had the same problem in a monorepo and have always hacked around with foreknowledge of where the repo would be at the time of a near-future release.  This is *much* cleaner and your implementation is spot-on. Thank you.

Please let me know if you have any questions about my comments and I look forward to a merge.

**Comment by cch1** [comment]
Also, you need to add a unit test for the `leiningen.v/dependency-version-from-scm` function before I can merge in good conscience.

File: project.clj (hunk 1 above)
Comment (chrisbetz) [reply project.clj]: Ok, that one turns out to be "necessary" because under Cursive/IntelliJ (and under some conditions in leiningen, e.g. when running "lein check") the project.clj file cannot be read with a keyword as "version". ("clojure.lang.Keyword cannot be cast to java.lang.String", see below)

But you're right, I just changed this ... but it's not essentially a part of the feature of this merge request.

So,  it's up to you to decide whether you want it or leave it.

Stack trace on `lein check`:

```
java.lang.ClassCastException: clojure.lang.Keyword cannot be cast to java.lang.String
 at leiningen.core.project$make_project_properties.invokeStatic (project.clj:19)
    leiningen.core.project$make_project_properties.invoke (project.clj:16)
    leiningen.core.eval$write_pom_properties.invokeStatic (eval.clj:41)
    leiningen.core.eval$write_pom_properties.invoke (eval.clj:36)
    leiningen.core.eval$prep.invokeStatic (eval.clj:84)
    leiningen.core.eval$prep.invoke (eval.clj:73)
    leiningen.core.eval$eval_in_project.invokeStatic (eval.clj:362)
    leiningen.core.eval$eval_in_project.invoke (eval.clj:356)
    leiningen.core.eval$eval_in_project.invokeStatic (eval.clj:360)
    leiningen.core.eval$eval_in_project.invoke (eval.clj:356)
    leiningen.check$check.invokeStatic (check.clj:32)
    leiningen.check$check.invoke (check.clj:8)
```

**Comment by chrisbetz** [comment]
Could you please give me a hint on how your tests should be run? Did not find and when running `lein test`, this is my output (i.e. no tests are executed, but I get some warnings):

```
❯ lein test                                                                                   master
fatal: bad revision 'HEAD'

No SCM data available!
fatal: bad revision 'HEAD'

Workspace is not anchored {:status {:tracking (## master...origin/master [ahead 1]), :files (A  x)}, :describe v2.3.4-alpha4-1-ge989-DIRTY}
Workspace is not anchored {:status {:tracking [## master...origin/master [ahead 1]], :files []}}
Workspace is not anchored {:status {:tracking [## master], :files [?? new-file.txt]}}

lein test integration.lein-v.plugin

lein test integration.leiningen.v

lein test leiningen.v.impl

lein test unit.leiningen.v

lein test unit.leiningen.v.file

lein test unit.leiningen.v.git

lein test unit.leiningen.v.maven

lein test unit.leiningen.v.semver

lein test utilities.git-shell

Ran 0 tests containing 0 assertions.
0 failures, 0 errors.
```


I'll happily add tests as soon as I get how you're doing it ;)

File: .gitignore [hunk 4]
```diff
@@ -11,3 +11,4 @@ pom.xml
 pom.xml.asc
 .lein-repl-history
 .nrepl-port
+/.idea/
```
[the remark this answers is not in the document]
Comment (chrisbetz) [reply .gitignore]: Will move it. To me, this has the same "level" as the .nrepl- stuff etc., but that's up to you :)

File: src/leiningen/v.clj (hunk 2 above)
Comment (chrisbetz) [reply src/leiningen/v.clj]: Will move that to the tests...

File: src/leiningen/v.clj [hunk 5]
```diff
@@ -103,15 +103,34 @@
...
+; ["abc" "1.0"]
+; => (update-dependency "42" ["abc" (quote lein-v)])
+; ["abc" "42"]
+
+
+(defn dependency-version-from-scm
+  [project]
+  (let [v (str (or (version (:v project)) "UNKNOWN"))
+        vk (str (or (:manifest-version-name (:v project)) "Implementation-Version"))]
+    (update-in project
```
[the remark this answers is not in the document]
Comment (chrisbetz) [reply src/leiningen/v.clj]: No particular reason (relic from copying assoc-in). Changed it to update...

File: src/leiningen/v.clj (hunk 3 above)
[the remark this answers is not in the document]
Comment (chrisbetz) [reply src/leiningen/v.clj]: consider it done :)

**Comment by chrisbetz** [comment]
Ah, forgot to mention that `lein midje` is not working also:

```
lein midje
'midje' is not a task.
```

Will try to find the proper "lein-midje" version next. It's been a while since I used midje... ;)

**Comment by chrisbetz** [comment]
Here you go. Got everything you requested (hopefully). Would love to see this merge :) Let me know if there's more to it.

Have a great day and thanks for your work and effort.

**Comment by cch1** [comment]
One more non-trivial request: please format your commit messages with one short summary line followed by more detail.  Ideally, following this approach: http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html

File: project.clj (hunk 1 above)
Comment (cch1) [reply project.clj]: I was unable to reproduce that failure... and then I remembered why.  In order to "eat our own dog food", I have a `.lein-classpath` file in the root of the project.  That _should_ allow `lein check` to work if you have a tolerably recent version of `lein-v` installed locally.

But it *must* be installed.  Can you try installing a recent version of `lein-v` locally and let me know if you can then run with the keyword?  I really would like to continue being "self aware" on this front.

Comment (chrisbetz) [reply project.clj]: oh, this does not help with IntelliJ, and I had lein-v installed locally... so, no, does not work for me. I'm using lein 2.7.1.

**Comment by chrisbetz** [comment]
Replaced by #10.

File: project.clj (hunk 1 above)
Comment (chrisbetz) [reply project.clj]: left this out of the PR #10.
